[JSC] Fix RISC-V64 compilation issues related to SIMD

Marie-Joseph opened the pull request:
#### 969f03eb1212b14808de814e57b1dfcaebc85293
<pre>
[JSC] Fix RISC-V64 compilation issues related to SIMD
<a href="https://bugs.webkit.org/show_bug.cgi?id=247488">https://bugs.webkit.org/show_bug.cgi?id=247488</a>

Reviewed by Žan Doberšek.

* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h: Implement vector instruction noops

Canonical link: <a href="https://commits.webkit.org/256872@main">https://commits.webkit.org/256872@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/264bf65624e71465ea3dea77d31203ae45374640

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96981 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6248 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30094 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106509 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166792 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6476 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34983 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89372 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103205 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102651 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4867 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83597 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31881 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86710 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88580 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74772 "Found 1 new API test failure: /WebKitGTK/TestWebKitUserContentFilterStore:/webkit/WebKitUserContentFilterStore/filter-persistence (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87918 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/282 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/83360 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/266 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21486 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28420 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4752 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5058 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43998 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/86055 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1498 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40788 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19393 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
This is not correct since this is breaking ARM64's loadVector etc. (these are defined in ARM64 too.)
Can you add these things into RISC-V instead?

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpersSpoolers.h:
```diff
@@ -272,7 +272,7 @@ class AssemblyHelpers::StoreRegSpooler : public AssemblyHelpers::Spooler<StoreRe
     ALWAYS_INLINE void executeVector(ptrdiff_t offset, FPRReg reg)
     {
-#if USE(JSVALUE64)
+#if USE(JSVALUE64) && !CPU(RISCV64)
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpersSpoolers.h:
```diff
@@ -211,7 +211,7 @@ class AssemblyHelpers::LoadRegSpooler : public AssemblyHelpers::Spooler<LoadRegS
     ALWAYS_INLINE void executeVector(ptrdiff_t offset, FPRReg reg)
     {
-#if USE(JSVALUE64)
+#if USE(JSVALUE64) && !CPU(RISCV64)
```
I think this is not correct. We should make it work.

Constellation reviewed (changes requested):
Commented

zdobersek commented on Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h:
```diff
@@ -1752,6 +1752,55 @@ class MacroAssemblerRISCV64 : public AbstractMacroAssembler<Assembler> {
+        UNUSED_PARAM(src);
+        UNUSED_PARAM(dest);
+        m_assembler.ebreakInsn();
+    }
```
There's the `MACRO_ASSEMBLER_RISCV64_TEMPLATED_NOOP_METHOD` macro that's usable for this purpose. It doesn't produce a breaking instruction, but it could. But it's useful to quickly cover different overloads for a given function that's not yet supported.

zdobersek commented on Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
```diff
@@ -743,7 +743,7 @@ MoveDouble U:F:64, D:F:64
     Addr, Addr, Tmp
 
-MoveVector U:F:128, D:F:128
+64: MoveVector U:F:128, D:F:128
```
These end up excluding RISCV64 until RISCV64 also starts being included under that condition since it's a 64-bit ISA. These can be avoided by using the macro to produce the necessary no-ops in the MacroAssemblerRISCV64 class.

zdobersek reviewed (approved):
💯

Commit:
[JSC] Fix RISC-V64 compilation issues related to SIMD
https://bugs.webkit.org/show_bug.cgi?id=247488

Reviewed by Žan Doberšek.

* Source/JavaScriptCore/assembler/MacroAssemblerRISCV64.h: Implement vector instruction noops

Canonical link: https://commits.webkit.org/256872@main
